### PR TITLE
[Sharktank][LLama][fp8] Add support for fp8 attn in Llama

### PR DESCRIPTION
**Pull request by dan-garvey**
```
 python3 -m sharktank.examples.export_paged_llm_v1 --irpa-file=/home/chi/src/test/llama/dan/fp8_attn.irpa \
--output-mlir=/home/chi/src/test/llama/dan/f8_attn_chi_castf32_roctorch.mlir \
--output-config=/home/chi/src/test/llama/dan/config_attn_chi.json \
--bs=1 --attention-kernel sharktank \
--attention-dtype=float8_e4m3fnuz --activation-dtype=bfloat16 --use-attention-mask --use-hf --kv-cache-dtype=float8_e4m3fnuz
```
**IRPA weight file**:
`sudo cp /home/dan/SHARK-Platform/fp8_attn.irpa /sharedfile/attn/fp8_attn.irpa`

**Export 0319 update**:
```
python3 -m sharktank.examples.export_paged_llm_v1 --irpa-file=/sharedfile/attn/fp8_attn.irpa \
--output-mlir=/sharedfile/attn/128/fp8_attn.mlir \
--output-config=/sharedfile/attn/128/config_attn.json \
--bs-prefill=4 --bs-decode=4 --attention-kernel sharktank \
--attention-dtype=float8_e4m3fnuz --activation-dtype=bfloat16 --use-attention-mask --use-hf --kv-cache-dtype=float8_e4m3fnuz
```

### CODE REVIEW

File: sharktank/sharktank/kernels/attention.py [hunk 1]
```diff
@@ -10,9 +10,110 @@
...
+        torch._check(q_d == k_d, lambda: f"expected matching qk features: {q_d}, {k_d}")
+
+        torch._check(k_s == v_s, lambda: f"expected matching kv length: {q_d}, {k_d}")
+
+        q_desc.specialize_dims(0, 1, -1)
+        k_desc.specialize_dims(0, 1, -1)
+        v_desc.specialize_dims(0, 1, -1)
+
+        # Result 0: Shape batch..., m, n
+        ksel.return_new_tensor((*q_bs, q_l, v_e), dtype=torch.float16).specialize_dims(
```
Comment (AmosLewis) [inline sharktank/sharktank/kernels/attention.py:66]:
```
        ksel.return_new_tensor((*q_bs, q_l, v_e), dtype=torch.float32).specialize_dims(
            0, 1, -1
        )
```
to fix error: `'tensor.cast' op operand type 'tensor<1x32x?x128xf32>' and result type 'tensor<1x32x?x128xf16>' are cast incompatible`

**Comment by AmosLewis** [comment]
Default bs1_input_32 iree compile and runs well without NAN. But I got iree-compile bug for bs4_input128. Should I file a iree-compile issue or it could be fixed here?
[llama_fp8_attn8_bs4_128_bug.txt](https://gist.github.com/AmosLewis/37d57aa044d727b56c442f11ed0ca02e#file-llama_fp8_attn8_bs4_bug-txt)
```
/sharedfile/attn/128/fp8_attn.mlir:29732:13: error: 'util.call' op function type mismatch; expected '(tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<f32>, tensor<?x?x?x?xf8E4M3FNUZ>) -> tensor<4x32x?x128xf32>' but callee is '(tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<f32>, tensor<?x?xf8E4M3FNUZ>) -> tensor<4x32x?x128xf32>'
    %2032 = "util.call"(%2026, %2027, %2028, %2031, %2030) <{callee = @sharktank_masked_flash_attention_4_32_128_128_f8E4M3FNUZ_f32_f32}> : (tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<f32>, tensor<?x?x?x?xf8E4M3FNUZ>) -> tensor<4x32x?x128xf32>
            ^
/sharedfile/attn/128/fp8_attn.mlir:29732:13: note: see current operation: %1905 = "util.call"(%1899, %1900, %1901, %1904, %1903) <{callee = @sharktank_masked_flash_attention_4_32_128_128_f8E4M3FNUZ_f32_f32}> : (tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<4x32x?x128xf8E4M3FNUZ>, tensor<f32>, tensor<?x?x?x?xf8E4M3FNUZ>) -> tensor<4x32x?x128xf32>
```

**Comment by AmosLewis** [comment]
Great, get benchmark for it. prefill_bs4_2048: 335ms. decode_bs4_2048:176ms, iree 3.3.0rc20250225 with pr https://github.com/iree-org/iree/pull/20091

**Comment by AmosLewis** [comment]
Since this patch rebased on TOM, I think we should add addition flag @sogartar  added `--kv-cache-dtype=float8_e4m3fnuz , --attention-dtype=bfloat16` https://github.com/nod-ai/shark-ai/pull/994#issuecomment-2679131469.
But we got [attn8_mask_export_bug.txt](https://gist.github.com/AmosLewis/a4f72fdf803fddf935bcda83815b63f0#file-attn8_mask_export_bug-txt) if added this flag in this pr.
```
python3 -m sharktank.examples.export_paged_llm_v1 --irpa-file=/sharedfile/llama3_8b_fp8.irpa  \
--output-mlir=/sharedfile/32/fp8_attn.mlir \
--output-config=/sharedfile/32/config_attn.json \
--bs=1 --attention-kernel sharktank \
--attention-dtype=bfloat16 \
--activation-dtype=bfloat16 \
--kv-cache-dtype=float8_e4m3fnuz \
--use-hf \
--use-attention-mask
...
  File "/home/chi/src/shark-ai/sharktank/sharktank/layers/paged_llama_attention_block.py", line 215, in forward
    assert self.attention_scale is not None
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```
Also try `--kv-cache-dtype=float8_e4m3fnuz , --attention-dtype=float8_e4m3fnuz `, failed with same assertion error. 
Why can't we set the `--kv-cache-dtype`? AFAIK, this patch's main contribution is enable use-attention-mask. And along with boian's previous achievement on kvcache and batchmatmul. It should give us better result. Is it something wrong in @dan-garvey patch or TOM kv-cache and matmul?

**Comment by dan-garvey** [comment]
```python3 -m sharktank.examples.export_paged_llm_v1 --irpa-file=fp8_attn.irpa --output-mlir=/home/dan/f8_attn_2.mlir --output-config=/home/dan/config.json --bs=4 --attention-kernel sharktank --attention-dtype=float8_e4m3fnuz --activation-dtype=bfloat16 --use-attention-mask --use-hf --kv-cache-dtype=float8_e4m3fnuz```

**Comment by IanWood1** [comment]
@AmosLewis fyi https://github.com/iree-org/iree/pull/20091 was merged

**Comment by AmosLewis** [comment]
Better to add whoever familiar with the algorithm to review it. When you are ready, I can update the benchmark.

**Comment by AmosLewis** [comment]
Could you share the ppl results in the comment right after the commit get it better? Also need the command you get the ppl so we can reproduce.

File: sharktank/tests/kernels/attention_template_test.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    def setUp(self):
+        torch.manual_seed(420)
+
+    @parameterized.expand(
+        [
+            (torch.float32, 5e-3, 1e-3, True),
+            (torch.float16, 5e-3, 1e-3, True),
+            # Currently failing on unmasked error
+            # (torch.float32, 5e-3, 1e-3, False),
+            # (torch.float16, 5e-3, 1e-3, False),
```
Comment (sogartar) [inline sharktank/tests/kernels/attention_template_test.py:33]: I think you could xfail parametrized tests.
https://stackoverflow.com/questions/30571481/how-to-mark-test-as-xfail-only-with-specific-parameters-with-pytest

File: sharktank/tests/kernels/attention_template_test.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        ref = torch.nn.functional.scaled_dot_product_attention(
+            q, k, v, mask, scale=scale
+        )
+
+        torch.testing.assert_close(res2.to(dtype), ref, atol=atol, rtol=rtol)
+
+    @parameterized.expand(
+        [
+            # Todo: fixed unmasked.
+            # (torch.float32, False, False),
```
Comment (sogartar) [inline sharktank/tests/kernels/attention_template_test.py:66]: You can xfail that.

File: sharktank/tests/kernels/attention_template_test.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            "masked_flash_attention"
+        )
+        cast = False
+        if dtype == torch.float8_e4m3fnuz:
+            dtype = torch.float32
+            cast = True
+        H = 4  # Head dim
+        N = 3  # Batch Size
+        L = 19  # Target Seq Len
+        S = 19  # Source Seq Len
```
Comment (sogartar) [inline sharktank/tests/kernels/attention_template_test.py:83]: Could we have different values for L, S to catch erroneous dimension mixing?

File: sharktank/tests/kernels/attention_template_test.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        ]
+    )
+    def test_export_dynamic(self, dtype, static, use_mask):
+        ops.attention_impls.register_attention_override_by_name(
+            "masked_flash_attention"
+        )
+        cast = False
+        if dtype == torch.float8_e4m3fnuz:
+            dtype = torch.float32
+            cast = True
```
Comment (sogartar) [inline sharktank/tests/kernels/attention_template_test.py:79]: If we make the op accept an output dtype we could simplify the user side.

File: sharktank/tests/kernels/attention_template_test.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        v = torch.rand([N, H, S, Ev], dtype=dtype)
+        if use_mask:
+            # mask is same type as inputs, therefore its added to score
+            mask = torch.rand([L, S], dtype=dtype)
+        if cast:
+            q = q.to(torch.float8_e4m3fnuz)
+            k = q.to(torch.float8_e4m3fnuz)
+            v = v.to(torch.float8_e4m3fnuz)
+            if use_mask:
+                mask = mask.to(torch.float8_e4m3fnuz)
```
Comment (sogartar) [inline sharktank/tests/kernels/attention_template_test.py:97]: I think in the real world we would never have a `float8_e4m3fnuz` mask as this type does not support infinites.
In the rest of the PR it seems you want the mask to be `f32`.

File: sharktank/tests/kernels/attention_template_test.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+        class MyModule(torch.nn.Module):
+            def forward(self, q, k, v, mask, scale):
+                return ops.scaled_dot_product_attention(
+                    q, k, v, a=mask, is_causal=False, scale=scale
+                )
+
+        mod = MyModule()
+        dtype = torch.dtype
+        if use_mask:
```
Comment (sogartar) [inline sharktank/tests/kernels/attention_template_test.py:121]: You could simplify this without branching if `mask` is set to `None` when `not use_mask` is true.

**Review by sogartar — APPROVED** [review]
My main concern is the hardcoding of dtypes. It seems that the attention operation could benefit from an output dtype argument.
If not in this PR, it would be nice to address this shortly.

**Comment by dan-garvey** [comment]
> Could you share the ppl results in the comment right after the commit get it better? Also need the command you get the ppl so we can reproduce.

It doesn't improve in iree due to discrepancies in the kernel. I don't plan to address ppl in this patch

**Comment by AmosLewis** [comment]
2048 decode benchmark drop from **35.1ms** to **20.5ms**. 128 not change.

File: sharktank/tests/kernels/attention_template_test.py (hunk 5 above)
Comment (dan-garvey) [reply sharktank/tests/kernels/attention_template_test.py:79]: this is actually just to create random tensors because you cant for fp8 in torch

File: sharktank/sharktank/layers/causal_llm.py [hunk 8]
```diff
@@ -96,7 +96,11 @@ def input_mask(
         return mask
 
     def decode_attention_mask(self, boolean_input_mask: torch.Tensor):
-        dtype = self.attention_dtype
+        dtype = (
+            torch.float32
+            if self.attention_dtype == torch.float8_e4m3fnuz
+            else self.attention_dtype
+        )
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/layers/causal_llm.py:103]: ideally eventually is_causal works and we dont pass a mask at all

File: sharktank/sharktank/models/llama/tools/import_quark_dataset.py [hunk 9]
```diff
@@ -218,10 +219,10 @@ def quantize_weight(
                 dtype=torch.float8_e4m3fnuz,
             )
         if output_quant_scale is not None:
-            updated_tensors[new_layer_name + ".qdq_output"] = StaticScaledQuantizer(
-                name=new_layer_name + ".qdq_output",
-                scale=1.0 / output_quant_scale,
-                reciprocal_scale=output_quant_scale,
+            updated_tensors[new_layer_name + ".q_output"] = StaticScaledQuantizer(
+                name=new_layer_name + ".q_output",
+                scale=1.0 / (output_quant_scale * 2.0),
```
[the remark this answers is not in the document]
Comment (dan-garvey) [reply sharktank/sharktank/models/llama/tools/import_quark_dataset.py:224]: the quantization is actually for e4m3fn but we have to convert to fnuz, which you can do by this multiplication. There is a comment elsehwere in this file, I can copy it here

**Review by IanWood1 — APPROVED** [review]

**Comment by AmosLewis** [comment]
https://github.com/nod-ai/shark-ai/pull/1098

**Comment by AmosLewis** [comment]
export faile @dan-garvey  [907_export_bug.log](https://gist.github.com/AmosLewis/252fdeaee1b279d6f92ce10ca8e581e7#file-907_export_bug-log)

**Comment by dan-garvey** [comment]
I know the fix, I'll push a pr and add an export test to data dependent tests

**Comment by AmosLewis** [comment]
New export llama f8 bug at the end of Mar 21 shark-ai https://gist.github.com/AmosLewis/252fdeaee1b279d6f92ce10ca8e581e7?permalink_comment_id=5508684#gistcomment-5508684

**Comment by AmosLewis** [comment]
> New export llama f8 bug at the end of Mar 21 shark-ai https://gist.github.com/AmosLewis/252fdeaee1b279d6f92ce10ca8e581e7?permalink_comment_id=5508684#gistcomment-5508684

2 follow up PR fix this export issue. 
#1152 
#1156
